File: src/main/java/mil/nga/tiff/internal/FileDirectoryRasterReader.java
package mil.nga.tiff.internal;

import mil.nga.tiff.field.FieldTypeDictionary;
import mil.nga.tiff.field.type.NumericFieldType;
import mil.nga.tiff.field.type.enumeration.PlanarConfiguration;
import mil.nga.tiff.field.type.enumeration.SampleFormat;
import mil.nga.tiff.internal.rasters.Rasters;
import mil.nga.tiff.io.ByteReader;
import mil.nga.tiff.util.TiffException;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.stream.IntStream;

public class FileDirectoryRasterReader {

    private final DirectoryStats stats;
    private final TileOrStripProcessor tileOrStripProcessor;
    private final FieldTypeDictionary typeDictionary;


    public FileDirectoryRasterReader(DirectoryStats stats, TileOrStripProcessor tileOrStripProcessor, FieldTypeDictionary typeDictionary) {
        this.stats = stats;
        this.tileOrStripProcessor = tileOrStripProcessor;
        this.typeDictionary = typeDictionary;
    }

    public Rasters readRasters(ImageWindow window, int[] samples, boolean sampleValues, boolean interleaveValues, ByteReader reader, boolean tiled) {

        // Validate the image window
        window.validate();
        window.validateFitsInImage(stats.imageWidth(), stats.imageHeight());

        int numPixels = window.numPixels();

        // Set or validate the samples
        int samplesPerPixel = stats.samplesPerPixel();
        if (samples == null) {
            samples = new int[samplesPerPixel];
            for (int i = 0; i < samples.length; i++) {
                samples[i] = i;
            }
        } else {
            for (int sample : samples) {
                if (sample >= samplesPerPixel) {
                    throw new TiffException("Invalid sample index: " + sample);
                }
            }
        }

        // Create the interleaved result buffer
        List<Integer> bitsPerSample = stats.bitsPerSample();
        int bytesPerPixel = 0;
        for (int i = 0; i < samplesPerPixel; ++i) {
            bytesPerPixel += bitsPerSample.get(i) / 8;
        }
        ByteBuffer interleave = null;
        if (interleaveValues) {
            interleave = ByteBuffer.allocateDirect(numPixels * bytesPerPixel);
            interleave.order(reader.getByteOrder());
        }

        // Create the sample indexed result buffer array
        ByteBuffer[] sample = null;
        if (sampleValues) {
            sample = new ByteBuffer[samplesPerPixel];
            for (int i = 0; i < sample.length; ++i) {
                double numberOfBytes = (double) numPixels * Double.valueOf(bitsPerSample.get(i)) / 8;

                if (numberOfBytes > Integer.MAX_VALUE) {
                    throw new TiffException("Number of sample value bytes is above max byte buffer capacity: " + numberOfBytes);
                }

                sample[i] = ByteBuffer.allocateDirect((int) numberOfBytes);
                sample[i].order(reader.getByteOrder());
            }
        }

        List<NumericFieldType> fieldTypes = IntStream.range(0, samples.length)
            .mapToObj(this::getFieldTypeForSample)
            .toList();

        // Create the rasters results
        Rasters rasters = new Rasters(window.width(), window.height(), fieldTypes, sample, interleave);

        // Read the rasters
        readRaster(window, samples, rasters, reader, tiled);

        return rasters;
    }

    /**
     * Read and populate the rasters
     *
     * @param window  image window
     * @param samples pixel samples to read
     * @param rasters rasters to populate
     */
    private void readRaster(ImageWindow window, int[] samples, Rasters rasters, ByteReader reader, boolean tiled) {

        int tileWidth = stats.tileWidth();
        int tileHeight = stats.tileHeight();

        int minXTile = window.minX() / tileWidth;
        int maxXTile = (window.maxX() + tileWidth - 1) / tileWidth;
        int minYTile = window.minY() / tileHeight;
        int maxYTile = (window.maxY() + tileHeight - 1) / tileHeight;

        int bytesPerPixel = getBytesPerPixel();

        int[] srcSampleOffsets = new int[samples.length];
        NumericFieldType[] sampleFieldTypes = new NumericFieldType[samples.length];
        for (int i = 0; i < samples.length; i++) {
            int sampleOffset = 0;
            if (stats.planarConfiguration() == PlanarConfiguration.CHUNKY) {
                sampleOffset = stats.bitsPerSample()
                        .subList(0, samples[i])
                        .stream()
                        .mapToInt(Integer::intValue)
                        .sum() / 8;
            }
            srcSampleOffsets[i] = sampleOffset;
            sampleFieldTypes[i] = getFieldTypeForSample(samples[i]);
        }

        for (int yTile = minYTile; yTile < maxYTile; yTile++) {
            for (int xTile = minXTile; xTile < maxXTile; xTile++) {

                int firstLine = yTile * tileHeight;
                int firstCol = xTile * tileWidth;
                int lastLine = (yTile + 1) * tileHeight;
                int lastCol = (xTile + 1) * tileWidth;

                for (int sampleIndex = 0; sampleIndex < samples.length; sampleIndex++) {
                    int sample = samples[sampleIndex];
                    if (stats.planarConfiguration() == PlanarConfiguration.PLANAR) {
                        bytesPerPixel = getSampleByteSize(sample);
                    }

                    byte[] block = tileOrStripProcessor.run(xTile, yTile, sample, reader, tiled, reader.getByteOrder());
                    ByteReader blockReader = new ByteReader(block, reader.getByteOrder());

                    for (int y = Math.max(0, window.minY() - firstLine); y < Math.min(tileHeight, tileHeight - (lastLine - window.maxY())); y++) {

                        for (int x = Math.max(0, window.minX() - firstCol); x < Math.min(tileWidth, tileWidth - (lastCol - window.maxX())); x++) {

                            int pixelOffset = (y * tileWidth + x) * bytesPerPixel;
                            int valueOffset = pixelOffset + srcSampleOffsets[sampleIndex];
                            blockReader.setNextByte(valueOffset);

                            // Read the value
                            Number value = sampleFieldTypes[sampleIndex].readValue(blockReader);

                            if (rasters.hasInterleaveValues()) {
                                int windowCoordinate = (y + firstLine - window.minY()) * window.width() + (x + firstCol - window.minX());
                                rasters.addToInterleave(sampleIndex, windowCoordinate, value);
                            }

                            if (rasters.hasSampleValues()) {
                                int windowCoordinate = (y + firstLine - window.minY()) * window.width() + x + firstCol - window.minX();
                                rasters.addToSample(sampleIndex, windowCoordinate, value);
                            }
                        }

                    }
                }
            }
        }
    }

    /**
     * Get the sample byte size
     *
     * @param sampleIndex sample index
     * @return byte size
     */
    private int getSampleByteSize(int sampleIndex) {
        List<Integer> bitsPerSample = stats.bitsPerSample();
        if (sampleIndex >= bitsPerSample.size()) {
            throw new TiffException("Sample index " + sampleIndex + " is out of range");
        }
        int bits = bitsPerSample.get(sampleIndex);
        if ((bits % 8) != 0) {
            throw new TiffException("Sample bit-imageWidth of " + bits + " is not supported");
        }
        return (bits / 8);
    }

    /**
     * Calculates the number of bytes for each pixel across all samples. Only
     * full bytes are supported, an exception is thrown when this is not the
     * case.
     *
     * @return the bytes per pixel
     */
    private int getBytesPerPixel() {
        int bitsPerSample = 0;
        List<Integer> bitsPerSamples = stats.bitsPerSample();
        for (int i = 0; i < bitsPerSamples.size(); i++) {
            int bits = bitsPerSamples.get(i);
            if ((bits % 8) != 0) {
                throw new TiffException("Sample bit-imageWidth of " + bits + " is not supported");
            } else if (bits != bitsPerSamples.getFirst()) {
                throw new TiffException("Differing size of samples in a pixel are not supported. sample 0 = " + bitsPerSamples.getFirst() + ", sample " + i + " = " + bits);
            }
            bitsPerSample += bits;
        }
        return bitsPerSample / 8;
    }

    /**
     * Get the field type for the sample
     *
     * @param sampleIndex sample index
     * @return field type
     */
    public NumericFieldType getFieldTypeForSample(int sampleIndex) {
        SampleFormat sampleFormat;
        List<SampleFormat> sampleFormatList = stats.sampleFormatList();
        if (sampleFormatList.isEmpty()) {
            sampleFormat = SampleFormat.UNSIGNED_INT;
        } else {
            int listId = sampleIndex < sampleFormatList.size() ? sampleIndex : 0;
            sampleFormat = sampleFormatList.get(listId);
        }
        int bitsPerSample = stats.bitsPerSample().get(sampleIndex);
        return typeDictionary.findBySampleParams(sampleFormat, bitsPerSample);
    }

}

File: src/main/java/mil/nga/tiff/internal/FileDirectoryRasterReader.kt
package mil.nga.tiff.internal

import mil.nga.tiff.field.TagDictionary
import mil.nga.tiff.field.type.NumericFieldType
import mil.nga.tiff.field.type.enumeration.PlanarConfiguration
import mil.nga.tiff.field.type.enumeration.SampleFormat
import mil.nga.tiff.internal.rasters.Rasters
import mil.nga.tiff.io.ByteReader
import mil.nga.tiff.util.TiffException
import java.nio.ByteBuffer
import java.util.stream.IntStream
import kotlin.math.max
import kotlin.math.min

class FileDirectoryRasterReader(
    private val stats: DirectoryStats,
    private val tileOrStripProcessor: TileOrStripProcessor,
    private val typeDictionary: TagDictionary,
    private val reader: ByteReader?
) {
    fun readRasters(
        window: ImageWindow,
        samples: IntArray?,
        sampleValues: Boolean,
        interleaveValues: Boolean,
        tiled: Boolean
    ): Rasters {

        reader!!

        // Validate the image window

        var actualSamples = samples
        window.validate()
        window.validateFitsInImage(stats.imageWidth!!, stats.imageHeight!!)

        val numPixels = window.numPixels()

        // Set or validate the samples
        val samplesPerPixel = stats.samplesPerPixel!!
        if (actualSamples == null) {
            actualSamples = IntArray(samplesPerPixel)
            for (i in actualSamples.indices) {
                actualSamples[i] = i
            }
        } else {
            for (sample in actualSamples) {
                if (sample >= samplesPerPixel) {
                    throw TiffException("Invalid sample index: $sample")
                }
            }
        }

        // Create the interleaved result buffer
        val bitsPerSample = stats.bitsPerSample!!
        var bytesPerPixel = 0
        for (i in 0..<samplesPerPixel) {
            bytesPerPixel += bitsPerSample[i] / 8
        }
        var interleave: ByteBuffer? = null
        if (interleaveValues) {
            interleave = ByteBuffer.allocateDirect(numPixels * bytesPerPixel)
            interleave.order(reader.byteOrder)
        }

        // Create the sample indexed result buffer array
        var sample: Array<ByteBuffer?>? = null
        if (sampleValues) {
            sample = arrayOfNulls(samplesPerPixel)
            for (i in sample.indices) {
                val numberOfBytes = numPixels.toDouble() * bitsPerSample[i] / 8

                if (numberOfBytes > Int.MAX_VALUE) {
                    throw TiffException("Number of sample value bytes is above max byte buffer capacity: $numberOfBytes")
                }

                sample[i] = ByteBuffer.allocateDirect(numberOfBytes.toInt())
                sample[i]!!.order(reader.byteOrder)
            }
        }

        val fieldTypes = IntStream.range(0, actualSamples.size)
            .mapToObj { sampleIndex: Int ->
                this.getFieldTypeForSample(
                    sampleIndex
                )
            }
            .toList()

        // Create the rasters results
        @Suppress("UNCHECKED_CAST") val rasters =
            Rasters.create(window.width(), window.height(), fieldTypes as List<NumericFieldType<*>>, sample?.requireNoNulls(), interleave)

        // Read the rasters
        readRaster(window, actualSamples, rasters, reader, tiled)

        return rasters
    }

    /**
     * Read and populate the rasters
     *
     * @param window  image window
     * @param samples pixel samples to read
     * @param rasters rasters to populate
     */
    private fun readRaster(
        window: ImageWindow,
        samples: IntArray,
        rasters: Rasters,
        reader: ByteReader,
        tiled: Boolean
    ) {
        val tileWidth = stats.tileWidth!!
        val tileHeight = stats.tileHeight!!

        val minXTile = window.minX / tileWidth
        val maxXTile = (window.maxX + tileWidth - 1) / tileWidth
        val minYTile = window.minY / tileHeight
        val maxYTile = (window.maxY + tileHeight - 1) / tileHeight

        var bytesPerPixel = bytesPerPixel

        val srcSampleOffsets = IntArray(samples.size)
        val sampleFieldTypesInit: Array<NumericFieldType<*>?> = arrayOfNulls(samples.size)
        for (i in samples.indices) {
            var sampleOffset = 0
            if (stats.planarConfiguration == PlanarConfiguration.CHUNKY) {
                sampleOffset = stats.bitsPerSample!!
                    .subList(0, samples[i])
                    .stream()
                    .mapToInt { obj: Int -> obj }
                    .sum() / 8
            }
            srcSampleOffsets[i] = sampleOffset
            sampleFieldTypesInit[i] = getFieldTypeForSample(samples[i])!!
        }

        val sampleFieldTypes = sampleFieldTypesInit.requireNoNulls()

        for (yTile in minYTile..<maxYTile) {
            for (xTile in minXTile..<maxXTile) {
                val firstLine = yTile * tileHeight
                val firstCol = xTile * tileWidth
                val lastLine = (yTile + 1) * tileHeight
                val lastCol = (xTile + 1) * tileWidth

                for (sampleIndex in samples.indices) {
                    val sample = samples[sampleIndex]
                    if (stats.planarConfiguration == PlanarConfiguration.PLANAR) {
                        bytesPerPixel = getSampleByteSize(sample)
                    }

                    val block = tileOrStripProcessor.run(xTile, yTile, sample, reader, tiled, reader.byteOrder)
                    val blockReader = ByteReader(block, reader.byteOrder)

                    val minY = max(0.0, (window.minY - firstLine).toDouble()).toInt()
                    val maxY = min(tileHeight.toDouble(), (tileHeight - (lastLine - window.maxY)).toDouble()).toInt()
                    for (y in minY..<maxY) {
                        val minX = max(0.0, (window.minX - firstCol).toDouble()).toInt()
                        val maxX = min(tileWidth.toDouble(), (tileWidth - (lastCol - window.maxX)).toDouble()).toInt()
                        for (x in minX..<maxX) {
                            val pixelOffset = (y * tileWidth + x) * bytesPerPixel
                            val valueOffset = pixelOffset + srcSampleOffsets[sampleIndex]
                            blockReader.nextByte = valueOffset

                            // Read the value
                            val value = sampleFieldTypes[sampleIndex].readValue(blockReader) as Number

                            if (rasters.hasInterleaveValues()) {
                                val windowCoordinate =
                                    (y + firstLine - window.minY) * window.width() + (x + firstCol - window.minX)
                                rasters.addToInterleave(sampleIndex, windowCoordinate, value)
                            }

                            if (rasters.hasSampleValues()) {
                                val windowCoordinate =
                                    (y + firstLine - window.minY) * window.width() + x + firstCol - window.minX
                                rasters.addToSample(sampleIndex, windowCoordinate, value)
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Get the sample byte size
     *
     * @param sampleIndex sample index
     * @return byte size
     */
    private fun getSampleByteSize(sampleIndex: Int): Int {
        val bitsPerSample = stats.bitsPerSample!!
        if (sampleIndex >= bitsPerSample.size) {
            throw TiffException("Sample index $sampleIndex is out of range")
        }
        val bits = bitsPerSample[sampleIndex]
        if ((bits % 8) != 0) {
            throw TiffException("Sample bit-imageWidth of $bits is not supported")
        }
        return (bits / 8)
    }

    private val bytesPerPixel: Int
        /**
         * Calculates the number of bytes for each pixel across all samples. Only
         * full bytes are supported, an exception is thrown when this is not the
         * case.
         *
         * @return the bytes per pixel
         */
        get() {
            var bitsPerSample = 0
            val bitsPerSamples = stats.bitsPerSample!!
            for (i in bitsPerSamples.indices) {
                val bits = bitsPerSamples[i]
                if ((bits % 8) != 0) {
                    throw TiffException("Sample bit-imageWidth of $bits is not supported")
                } else if (bits != bitsPerSamples.first()) {
                    throw TiffException("Differing size of samples in a pixel are not supported. sample 0 = " + bitsPerSamples.first() + ", sample " + i + " = " + bits)
                }
                bitsPerSample += bits
            }
            return bitsPerSample / 8
        }

    /**
     * Get the field type for the sample
     *
     * @param sampleIndex sample index
     * @return field type
     */
    fun getFieldTypeForSample(sampleIndex: Int): NumericFieldType<*>? {
        val sampleFormat: SampleFormat
        val sampleFormatList = stats.sampleFormatList!!
        if (sampleFormatList.isEmpty()) {
            sampleFormat = SampleFormat.UNSIGNED_INT
        } else {
            val listId = if (sampleIndex < sampleFormatList.size) sampleIndex else 0
            sampleFormat = sampleFormatList[listId]
        }
        val bitsPerSample = stats.bitsPerSample!![sampleIndex]
        return typeDictionary.findNumericTypeBySampleParams(sampleFormat, bitsPerSample)
    }
}
